runtime/vm/CRIUHelpers.cpp

```diff
@@ -425,12 +434,96 @@ criuRestoreInitializeDump(J9VMThread *currentThread, void *userData, const char
 				*nlsMsgFormat = j9nls_lookup_message(J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE,
 						J9NLS_VM_CRIU_RESTORE_INITIALIZE_DUMP_FAILED, NULL);
 				result = FALSE;
+			} else {
+				vm->checkpointState.flags |=  J9VM_CRIU_TRANSITION_TO_DEBUG_INTERPRETER;
 			}
 		}
 	}
 	return result;
 }
 
+/**
+ * An internal JVM checkpoint hook to check for a debug interpreter request after restore.
+ *
+ * @param[in] currentThread vmThread token
+ * @param[in] userData J9InternalHookRecord pointer
+ * @param[in/out] nlsMsgFormat an NLS message
+ *
+ * @return BOOLEAN TRUE
+ */
+static BOOLEAN
+criuRestoreCheckForDebugInterpreterRequest(J9VMThread *currentThread, void *userData, const char **nlsMsgFormat)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+
+	if (NULL != vm->checkpointState.restoreArgsList) {
+		J9VMInitArgs *restoreArgsList = vm->checkpointState.restoreArgsList;
+		IDATA debugOn = FIND_AND_CONSUME_ARG(restoreArgsList, EXACT_MATCH, VMOPT_XXDEBUGINTERPRETER, NULL);
+		IDATA debugOff = FIND_AND_CONSUME_ARG(restoreArgsList, EXACT_MATCH, VMOPT_XXNODEBUGINTERPRETER, NULL);
+		if (debugOn > debugOff) {
+			vm->checkpointState.flags |= J9VM_CRIU_TRANSITION_TO_DEBUG_INTERPRETER;
+		}
+	}
+
+	return TRUE;
+}
+
+/**
+ * An internal JVM checkpoint hook to check for a JDWP request after restore.
+ *
+ * @param[in] currentThread vmThread token
+ * @param[in] userData J9InternalHookRecord pointer
+ * @param[in/out] nlsMsgFormat an NLS message
+ *
+ * @return BOOLEAN TRUE
+ */
+static BOOLEAN
+criuRestoreCheckForJdwp(J9VMThread *currentThread, void *userData, const char **nlsMsgFormat)
+{
+#define AGENTPATH_JDWP VMOPT_AGENTPATH_COLON "jdwp"
+#define LIB_JDWP "libjdwp.so"
+	J9JavaVM *vm = currentThread->javaVM;
+
+	if (NULL != vm->checkpointState.restoreArgsList) {
+		J9VMInitArgs *restoreArgsList = vm->checkpointState.restoreArgsList;
+		bool foundJdwp = false;
+		if ((FIND_ARG_IN_ARGS(restoreArgsList, STARTSWITH_MATCH, MAPOPT_AGENTLIB_JDWP_EQUALS, NULL) >= 0)
+			|| (FIND_ARG_IN_ARGS(restoreArgsList, STARTSWITH_MATCH, MAPOPT_XRUNJDWP, NULL) >= 0)
+			|| (FIND_ARG_IN_ARGS(restoreArgsList, STARTSWITH_MATCH, AGENTPATH_JDWP, NULL) >= 0)
+		) {
+			foundJdwp = true;
+		} else {
+			IDATA agentPathIndex = FIND_ARG_IN_ARGS(restoreArgsList, STARTSWITH_MATCH, VMOPT_AGENTPATH_COLON, NULL);
+			if (agentPathIndex >= 0) {
+				char *agentPathValue = NULL;
+				GET_OPTION_VALUE_ARGS(restoreArgsList, agentPathIndex, ':', &agentPathValue);
+				if (NULL != agentPathValue) {
+					UDATA libraryLength;
+					char *option = strstr(agentPathValue, "=");
+					if (NULL == option) {
+						libraryLength = strlen(agentPathValue);
+					} else {
+						libraryLength = option-agentPathValue;
+					}
+
+					if (0 == strncmp(LIB_JDWP, agentPathValue + libraryLength - LITERAL_STRLEN(LIB_JDWP), LITERAL_STRLEN(LIB_JDWP))) {
+						foundJdwp = true;
+					}
+				}
+			}
+		}
+
+		if (foundJdwp) {
```

Comment:
I think this `criuRestoreCheckForJdwp ()` searching for the `JDWP` agent can be replaced by just checking `J9VM_CRIU_IS_JDWP_ENABLED`.
`checkTransitionToDebugInterpreter()` now is moved after `runInternalJVMRestoreHooks()`, `TRIGGER_J9HOOK_VM_PREPARING_FOR_RESTORE()` will invoke `criuRestoreInitializeLib()` which sets `J9VM_CRIU_IS_JDWP_ENABLED` accordingly.

Reply:
Ok. Good point.